### PR TITLE
#1404 Serial Monitor message history

### LINKED ISSUE #1404: Serial Monitor message history
### Describe the request

In Arduino IDE 1.x, you can 'up arrow' through the recently sent commands, to send the same one a second time, or even go further back in history.

### Describe the current behavior

none, this functionality is missing.

### Arduino IDE version

IDE 2.0.0-rc9.3

### Operating system

macOS

### Operating system version

Monterey

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### PR DESCRIPTION
**Pull request by dwightfowler-cd**
### Motivation
Address PR #1453 comment

### Change description
Conform to community coding guidelines

### Other information
Closes #1404

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by dwightfowler-cd** [comment]
@kittaakos, is everything go to go for this PR? Is there anything else I should do to get this ready?

**Comment by nmzaheer** [comment]
I have reviewed the code and have some queries:

1. What happens when I press the Up Arrow to retrieve the previous command and then press Enter? Does it get added to the list or since it is a duplicate do you drop it?
2. The legacy code also had a provision of handling Escape key to reset? Are we dropping it? Or is it handled internally?

I will manually test your code out and give you feedback maybe tomorrow

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 1]
```diff
@@ -6,6 +6,74 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
 import { MonitorModel } from '../../monitor-model';
 import { Unknown } from '../../../common/nls';
 
+class RingList {
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:9]: `RingList` implies that message history is a cycle.

With the following sent messages:
 - `one`,
 - `two`, and
 - `three`.

I would expect to get `three` after `one`. But no, so the class name is wrong.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 2]
```diff
@@ -6,6 +6,74 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
...
+  private ring: string[];
+  private size: number;
+  private begin: number;
+  private index: number;
+  private end: number;
+
+  constructor(size: number = 100) {
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:16]: Should be `constructor(size = 100) {`.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 3]
```diff
@@ -6,6 +6,74 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
...
+    this.init = this.init.bind(this);
+    this.push = this.push.bind(this);
+    this.prev = this.prev.bind(this);
+    this.next = this.next.bind(this);
+    this.init(size);
+  }
+
+  private init(size: number = 100)
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:24]: Should be `private init(size = 100) {`

Please respect the position of the curly braces.

**Review by kittaakos — CHANGES_REQUESTED** [review]
I gave it a try, but it behaves differently than in IDE 1.x.

Steps to reproduce:
 - send `one`,
 - send `two`,
 - press up -> the history must be `two`, but it's `one`

If you fix this, I take another look.

IDE 1.x:



https://user-images.githubusercontent.com/1405703/192241337-1c2367af-ac16-4758-a7e8-a774355b21f9.mp4


IDE2:

https://user-images.githubusercontent.com/1405703/192241257-cfcabc40-e9ac-42c1-9fc4-82b7f965cf73.mp4

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 1 above)
Comment (dwightfowler-cd) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:9]: True. I debated with myself over that name. I'm changing it to HistoryStack. It's more descriptive.

**Comment by nmzaheer** [comment]
It's almost been 3 weeks since the last update. How can I contribute and complete the required changes so that this feature is completed?

**Comment by kittaakos** [comment]
Thank you for your initial work on this feature, @dwightfowler-cd.

There is another PR: #1562 based on your contribution. Please follow that one.

I am closing this as a duplicate.

**Pull request closed without merging**